[custom_ops] make custom_ops.impl work on existing operators

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #106088
* __->__ #106076
* #106075

The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

zou3519 committed:
[custom_ops] make custom_ops.impl work on existing operators

The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] make custom_ops.impl work on existing operators"


The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] make custom_ops.impl work on existing operators"


The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

[ghstack-poisoned]

soulitzer commented:
> If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

This sounds good, but just curious - are there plans to also open it up to allow users to override existing implementations? Perhaps with an `override=True` flag? Or is that something where we tell users you should just use the lower-level torch library API instead.

soulitzer commented on torch/_custom_ops.py:
```diff
@@ -169,7 +169,7 @@ def impl(qualname, *, device_types=("cpu", "cuda"), func=None):
 
     def inner(func):
-        custom_op = _find_custom_op(qualname)
+        custom_op = _find_custom_op(qualname, also_check_torch_library=True)
```
Maybe we want to advertise this new feature on the doc string?

zou3519 replied:
We could!

zou3519 commented:
> This sounds good, but just curious - are there plans to also open it up to allow users to override existing implementations? Perhaps with an override=True flag? Or is that something where we tell users you should just use the lower-level torch library API instead.

override=True sounds reasonable to me

zou3519 commented:
@soulitzer I am going to merge this first without the override= option. The implementation warrants an additional PR, plus I want to see how people use this first without overide=.

zou3519 committed:
Update on "[custom_ops] make custom_ops.impl work on existing operators"


The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] make custom_ops.impl work on existing operators"


The design is that we construct a CustomOp object around the existing
operator and then use it to register things. It is totally OK if the
operator isn't functional (unlike torch._custom_ops.custom_op that can
only construct functional operators).

If the operator already has an implementation from a backend (either via
direct registration to e.g. DispatchKey::CPU, or an indirect
registration like CompositeImplicitAutograd/CompositeExplicitAutograd),
we raise an error.

Test Plan:
- new tests

[ghstack-poisoned]
